### PR TITLE
added e2e test for avocode email tags input

### PR DESCRIPTION
**Pull request by Talank**
### Description
 Some scenarios for E2E test of basic and unique avocode-email-tags-input has been added

### Related Issue
https://github.com/avocode/avocode-email-tagsinput/issues/19

### CODE REVIEW

File: packages/avocode-email-tagsinput/test/e2e/basic.spec.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    ]
+    const supportedEmailEdgeCases = [
+      '1234@test.co',
+      'testemail@11.ex',
+      "single'123@quote.co",
+      'withcu{}ly@bracket.co',
+      'सिमप्लेफोल्देर$%#?&@nepali.co',
+    ]
+
+    for (const testEmail of unsupportedEmailEdgeCases) {
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/basic.spec.js:87]: Please do not create dynamic tests. I know you mean well and I should probably create a contribution guide or something but I am strongly opposed to these.

My reasons are that they don't really bring much value because the "spec" should be "specification" of test and clearly state what you are testing.

They are also harder to debug especially when they're implemented in the loops since adding a debugger now means stepping through other specs. Basically when test fails, I want to be able to jump to that specific failing spec and fix it.

Also the description does not really say why I am not allowed to add specific tag. There should be simply a spec like this:

`should not be allowed to add invalid email as a tag`

You do not need to test every combination of invalid emails, I don't see a point in it.

File: packages/avocode-email-tagsinput/test/e2e/basic.spec.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    it('should navigate one char left', async () => {
+      await page.click(tagsInputFieldSelector)
+      await page.type(tagsInputFieldSelector, 'abcd')
+      await page.keyboard.press('ArrowLeft')
+      const [ anchorOffset ] = await getAnchorAndFocus(page)
+      expect(anchorOffset).to.equal(3)
+    })
+  })
+
+  describe('user typing different email edge cases', () => {
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/basic.spec.js:70]: I think this whole `describe` block should be deleted. You can just have two specs testing valid email and invalid email and that's it.

The logic behind validating email address is implemented in `isValueValidEmail` and you should probably write unit tests for utils. It's a different concern.

File: packages/avocode-email-tagsinput/test/e2e/basic.spec.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      })
+    }
+
+    it('spaces provided inside email are omitted while creating email tag', async () => {
+      await addEmailTag(page, 'em a il@s pa ce. co')
+      await assertInputText(page, 'email@space.co')
+    })
+  })
+
+  describe('user can add new tag typing valid email and hitting comma, enter or space', () => {
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/basic.spec.js:107]: OK this is similar point to dynamic specs above so the same applies here.

It's even more evident here. It'd be better to be verbose here, I know you're trying to do things "smart" here but my opinion is it's better to be verbose in specs (from my experience).

I think it's better to have a specific spec saying `it should be possible to ad dan email tag hitting comma key` and then test for that specific key.

File: packages/avocode-email-tagsinput/test/e2e/test-utils.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  if (tagSubmitKeys.includes(submitKey)) {
+    await page.keyboard.press(submitKey)
+  } else {
+    throw new Error(`invalid submit key provided. Expected ${
+      tagSubmitKeys.toString()}, but found: ${submitKey}`
+    )
+  }
+}
+
+const assertEmailTagCount = async function (page, expectedCount) {
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/test-utils.js:56]: I'm sorry, I know you're trying to save yourself typing the same code over but I really dislike these "assert*" helper methods. I think in the long run they will be more trouble than help.

I think they are hard to read inside the tests. Because when specific test fails, I'm most likely run into the helper method and then I have to go looking into the helper method finding... two lines of code.

Also the `assert...` is weird, I mean the naming is good it's just I'm using `chai` and `jest` here and the naming is to have `expect` calls in the specs. That way I can easily see where the condition is being tested.

So please remove these and just add the code inline. The other helper methods are actually fine and I'm glad you re-used the pattern of testing utilities from `better-react-tagsinput` package 👍

File: packages/avocode-email-tagsinput/test/e2e/unique.spec.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+  describe('user typing different email edge cases', () => {
+    const uniqueEmailExamples = [
+      "single'123@quote.aa",
+      'withcu{}ly@bracket.go',
+    ]
+
+    it('should be allowed to add unique email tags', async () => {
+      let tagCount = 0
+      for (const testEmail of uniqueEmailExamples) {
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/unique.spec.js:44]: Delete dynamic specs as explained in other comments. Just add a single spec where you're adding unique emails and that's it.

File: packages/avocode-email-tagsinput/test/e2e/unique.spec.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        await addEmailTag(page, testEmail)
+        await assertEmailTagCount(page, ++tagCount)
+      }
+    })
+
+    it('should not be allowed to add same email tag again', async () => {
+      await addEmailTag(page, uniqueEmailExamples[0])
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/unique.spec.js:51]: Please be more verbose here, don't use `uniqueEmailExamples` variable.

I actually think this spec is hard to read. It'd be more clear to have test specify specify email and then reading below that you're trying to add the same email.

File: packages/avocode-email-tagsinput/test/e2e/unique.spec.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  })
+
+  it('should render editor', async () => {
+    const editor = await page.waitForSelector(tagsInputFieldSelector)
+    expect(editor).to.exist
+  })
+
+  describe('user typing different email edge cases', () => {
+    const uniqueEmailExamples = [
+      "single'123@quote.aa",
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/unique.spec.js:38]: When using testing emails, please use some generic ones, these look spammy and I'm not sure about them. Why not `foo@bar.com` or something like that?

File: packages/avocode-email-tagsinput/test/e2e/basic.spec.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
+import puppeteer from 'puppeteer'
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/basic.spec.js:1]: Please rename this file to be `packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js` since you're directly testing that single component.

I think we haven't got to the point where we should separate spec files based on testing scenarios. So let's keep it in one file for now.

File: packages/avocode-email-tagsinput/test/e2e/unique.spec.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
+import puppeteer from 'puppeteer'
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/unique.spec.js:1]: Delete this file and instead put the specs in new `describe` block where you actually describe that you're testing uniqueness. I don't think we need separate file for this for now.

File: packages/avocode-email-tagsinput/test/e2e/basic.spec.js (hunk 2 above)
Comment (comatory) [reply packages/avocode-email-tagsinput/test/e2e/basic.spec.js:70]: What I mean is that there should be unit test testing file `lib/utils/index` where you would write unit tests for testing input / output of these utils.

In E2E tests I think it's OK to test for valid / invalid cases. (We're not writing [smoke tests](https://en.wikipedia.org/wiki/Smoke_testing_%28software%29) here so I think we're safe).

**Comment by Talank** [comment]
@comatory Thankyou for the review. 
I have merged the two files `basic..` and `unique..` into one as you have suggested. Also, I have removed the dynamic test and `assert..` method from testUtils. 
However, I have noticed random fails on Travis build! :thinking:
Tell me if I need to change anything.

File: packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    await page.close()
+  })
+
+  afterAll(async () => {
+    await browser.close()
+  })
+
+  const navigateToUnique = async () => {
+    const elements = await page.$x('//a[@href="#/avocode-email-tagsinput/unique"]')
+    await elements[0].click()
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js:34]: This is not nice modifying DOM anyway. You'd want to do redirect. 

Also this is what I mean in my previous review when I said it'd be better to create special testing routes in `playground` package just for testing purposes. This is just unnecessary complexity. Don't do it in this PR though, just a food for thought how to do it next time.

**Comment by comatory** [comment]
> @comatory Thankyou for the review.
> I have merged the two files `basic..` and `unique..` into one as you have suggested. Also, I have removed the dynamic test and `assert..` method from testUtils.
> However, I have noticed random fails on Travis build! 🤔
> Tell me if I need to change anything.

It's the spec you uncommented in https://github.com/avocode/avocode-email-tagsinput/pull/16

```
FAIL ../better-react-tagsinput/test/e2e/collapsible-tagsinput.spec.js (18.051s)

  ● CollapsibleTagsinput › counter › multiple tags › should update count when tags are added and input is on multiple lines

    Expected value   3

    Received:

      0

    Message:

      expected 0 to equal 3

      343 |         })

      344 | 

    > 345 |         expect(count).to.equal(3)

          |                          ^

      346 |       })

      347 |     })

      348 |   })

      at Object.<anonymous> (packages/better-react-tagsinput/test/e2e/collapsible-tagsinput.spec.js:345:26)
```

It has to be "fixed". So there was a reason after all why it was skipped. Perhaps it's a timing issue, I'm not sure.

I'll probably revert your PR and deal with this myself some other time. Let's just focus on this PR now.

**Comment by comatory** [comment]
OK, the bad spec will be dealt with in another PR. I might look into it and fix it, I don't think it's a big deal but unskipping it I didn't realize that this happened randomly.

Anyway rebase your branch with `master` and the checks should pass.

**Comment by comatory** [comment]
Also please squash your commits into one commit, I don't think we need to have the review process recorded in git history.

**Comment by Talank** [comment]
@comatory I made the suggested changes. Let me know if further changes are required.
I will squash all the commit before merge, if everything is ok.

File: packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js [hunk 11]
```diff
@@ -0,0 +1,154 @@
+import puppeteer from 'puppeteer'
+import { expect } from 'chai'
+import {
+  URLRoot,
+  tagsInputFieldSelector,
+  getAnchorAndFocus,
+  addEmailTag,
+  getTagNodes, getInputText,
+} from './test-utils'
+describe('Avocode Email Tags Input', () => {
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js:10]: 1. Please respect meaningful whitespace here. Put new line between imports and describe block, this looks really ugly.
2. Let's keep this consistent with `better-react-tagsinput` library, let's just add name of the component here: `AvocodeEmailTagsinput`

File: packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+
+  describe('Basic Email Tags Input', () => {
+    beforeEach(async () => {
+      page = await browser.newPage()
+      await page.goto(URLRoot)
+    })
+
+    it('should render editor', async () => {
+      const editor = await page.waitForSelector(tagsInputFieldSelector)
+      expect(editor).to.exist
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js:34]: I hate to be a strickler, but I'd kindly ask you to add new line between spec `expect` and rest of the code to make it consistent with other specs in `better-react-tagsinput`.

I should probably add linting rules to enforce this.

**Comment by Talank** [comment]
> > @comatory Thankyou for the review.
> > I have merged the two files `basic..` and `unique..` into one as you have suggested. Also, I have removed the dynamic test and `assert..` method from testUtils.
> > However, I have noticed random fails on Travis build! thinking
> > Tell me if I need to change anything.
> 
> It's the spec you uncommented in #16
> 
> ```
> FAIL ../better-react-tagsinput/test/e2e/collapsible-tagsinput.spec.js (18.051s)
> 
>   ● CollapsibleTagsinput › counter › multiple tags › should update count when tags are added and input is on multiple lines
> 
>     Expected value   3
> 
>     Received:
> 
>       0
> 
>     Message:
> 
>       expected 0 to equal 3
> 
>       343 |         })
> 
>       344 | 
> 
>     > 345 |         expect(count).to.equal(3)
> 
>           |                          ^
> 
>       346 |       })
> 
>       347 |     })
> 
>       348 |   })
> 
>       at Object.<anonymous> (packages/better-react-tagsinput/test/e2e/collapsible-tagsinput.spec.js:345:26)
> ```
> 
> It has to be "fixed". So there was a reason after all why it was skipped. Perhaps it's a timing issue, I'm not sure.

I also think it is the timing issue. However, it is not the only test that is failing each time. For example, in [this build](https://travis-ci.org/github/avocode/avocode-email-tagsinput/builds/677652473), some other test is failing.

> I'll probably revert your PR and deal with this myself some other time. Let's just focus on this PR now.

File: packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    })
+
+    describe('user typing', () => {
+      it('should insert text', async () => {
+        await page.click(tagsInputFieldSelector)
+        await page.type(tagsInputFieldSelector, 'abcd')
+        const text = await getInputText(page)
+        expect(text)
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js:42]: No idea why you changed the whitespace here. `expect(text).to.equal('abcd')` reads way better for me than this.

I think it's OK to do this once you have really long stiff in `equal` expressions but it's not required here.

File: packages/avocode-email-tagsinput/test/e2e/test-utils.js [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    const selection = window.getSelection()
+
+    return [ selection.anchorOffset, selection.focusOffset ]
+  })
+}
+
+const addEmailTag = async (page, emailTag, submitKey = 'Comma') => {
+  await page.click(tagsInputFieldSelector)
+  await page.type(tagsInputFieldSelector, emailTag)
+  if (tagSubmitKeys.includes(submitKey)) {
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/test-utils.js:45]: Not sure why you would do this kind of check here. It's not a concern of testing utility to check for "valid" keys because this library does not have concept of valid keys. You can use `addTagKeyCodes` to pass any key that will be used for adding tags, yes even nonsensical but I saw no point in enforcing it.

Comma, enter and space are just meaningful defaults that can be overriden. Also there's Flow type checker that will check for numeric type so no point in this really.

File: packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+        await addEmailTag(page, validEmailTag, 'Space')
+        const tagNodes = await getTagNodes(page)
+        expect(tagNodes)
+          .to
+          .have
+          .length(1)
+      })
+    })
+  })
+  describe('Unique Email Tags Input', () => {
```
Comment (comatory) [inline packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js:128]: Please add whitespace here as well.

Comment (comatory) [reply packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js:128]: And pleas change description to `unique emails`

**Comment by comatory** [comment]
As for the randomly failing specs, I'll have to look into it. Originally, these specs ran on our internal infrastructure where we maintained our own Docker containers so not sure if my migration to Github is not correctly done. I'll keep an eye out. My next step will probably be setting up Travis locally and running tests over and over again and see if I can debug it.

File: packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js (hunk 15 above)
Comment (Talank) [reply packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js:127]: I will fix those white spaces. I wonder how those spaces appeared! Maybe I accidentally gave the command in my IDE. Sorry about that.

Comment (comatory) [reply packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js:127]: Maybe you use Prettier or some auto-fix.

**Comment by Talank** [comment]
ok, I think all spaces fixed now. Also, I modified the code so that it is not checking the value in the submit key. I will squash commits if it looks good to you.

File: packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js (hunk 11 above)
Comment (comatory) [reply packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js:11]: Can you fix this please?

File: packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js (hunk 12 above)
Comment (comatory) [reply packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js:35]: Can you please add new line everywhere, so it looks like this?

```js
const editor = await page.waitForSelector(tagsInputFieldSelector)

expect(editor).to.exist 
```

Please make it analogically everywhere.

File: packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js (hunk 15 above)
Comment (comatory) [reply packages/avocode-email-tagsinput/test/e2e/avocode-email-tagsinput.spec.js:108]: Please add new line above `describe`

**Comment by Talank** [comment]
Ok cool :partying_face: 
Thanks for the quick interaction.
